File: week15/week 15-5.cpp

int equalSubstring(char* s, char* t, int maxCost) {
    int N = strlen(s); //先知道字串長度
    int ans = 0, j = 0;

    for(int i=0; i<N; i++){ //字串的迴圈
        maxCost -=abs(s[i] - t[i]);
        while(maxCost<0){
            maxCost += abs(s[j] - t[j]);
            j +=1;
       }
        if (i-j+1>ans) ans = i-j+1;
    }
    return ans; //課堂作業一 先隨便給個0當答案(答案不對')
}

File: week15/week 15-5.py
class Solution:
    def equalSubstring(self, s: str, t: str, maxCost: int) -> int:
        ans = 0
        N = len(s)
        j = 0 #尾巴
        for i in range(N): #頭是i 先用python
            maxCost-=abs(ord(s[i]) - ord(t[i]))
            while maxCost <0: #預算不夠 變負的
                maxCost += abs(ord(s[j]) - ord(t[j]))
                j += 1 #尾巴j 往右縮
            ans = max(ans, i-j+1) #頭i -

        return ans
